File: test_connection.py
"""Test script to verify Polymarket connection and configuration."""

import sys
from config import Config
from logger import logger


def test_configuration():
    """Test configuration loading."""
    print("\n" + "="*60)
    print("CONFIGURATION TEST")
    print("="*60)

    try:
        Config.validate()
        print("[OK] Configuration loaded successfully")
        print(f"  - Chain ID: {Config.CHAIN_ID}")
        print(f"  - Signature Type: {Config.SIGNATURE_TYPE}")
        print(f"  - Order Size: ${Config.ORDER_SIZE_USD}")
        print(f"  - Spread Offset: {Config.SPREAD_OFFSET}")
        print(f"  - Check Interval: {Config.CHECK_INTERVAL_SECONDS}s")
        return True
    except Exception as e:
        print(f"[FAIL] Configuration error: {e}")
        return False


def test_gamma_api():
    """Test connection to Gamma API."""
    print("\n" + "="*60)
    print("GAMMA API TEST")
    print("="*60)

    try:
        from market_discovery import MarketDiscovery

        discovery = MarketDiscovery()
        print("[OK] Market discovery client initialized")

        print("  Fetching markets...")
        markets = discovery.discover_btc_15m_markets()

        print(f"[OK] Successfully connected to Gamma API")
        print(f"  - Found {len(markets)} BTC 15m markets")

        if markets:
            print("\n  Recent markets:")
            for market in markets[:3]:
                print(f"    - {market.market_slug}")
                print(f"      Start: {market.start_datetime}")

        return True

    except Exception as e:
        print(f"[FAIL] Gamma API error: {e}")
        logger.error("Gamma API test failed", exc_info=True)
        return False


def test_clob_client():
    """Test CLOB client initialization."""
    print("\n" + "="*60)
    print("CLOB CLIENT TEST")
    print("="*60)

    try:
        from order_manager import OrderManager

        print("  Initializing CLOB client...")
        manager = OrderManager(Config.PRIVATE_KEY)

        print(f"[OK] CLOB client initialized")
        print(f"  - Wallet address: {manager.address}")

        print("  Fetching USDC balance...")
        balance = manager.get_usdc_balance()

        print(f"[OK] Successfully connected to CLOB API")
        print(f"  - USDC Balance: ${balance:.2f}")

        if balance < Config.ORDER_SIZE_USD * 4:
            print(f"\n  [WARNING] Low balance!")
            print(f"    You need at least ${Config.ORDER_SIZE_USD * 4:.2f} USDC")
            print(f"    to place all orders (4 orders × ${Config.ORDER_SIZE_USD})")

        return True

    except Exception as e:
        print(f"[FAIL] CLOB client error: {e}")
        logger.error("CLOB client test failed", exc_info=True)
        return False


def main():
    """Run all tests."""
    print("\n" + "="*60)
    print("   Polymarket Limit Order Bot - Connection Test")
    print("="*60)

    results = []

    # Test configuration
    results.append(("Configuration", test_configuration()))

    # Test Gamma API
    results.append(("Gamma API", test_gamma_api()))

    # Test CLOB client
    results.append(("CLOB Client", test_clob_client()))

    # Summary
    print("\n" + "="*60)
    print("TEST SUMMARY")
    print("="*60)

    all_passed = True
    for name, passed in results:
        status = "[PASSED]" if passed else "[FAILED]"
        print(f"{name:20s} {status}")
        if not passed:
            all_passed = False

    print("="*60)

    if all_passed:
        print("\n*** All tests passed! Your bot is ready to run.")
        print("\nTo start the bot:")
        print("  py main.py")
        print("\nOr use the run script:")
        print("  Windows: run.bat")
        print("  Unix/Mac: ./run.sh")
        return 0
    else:
        print("\n*** Some tests failed. Please fix the errors above.")
        print("\nCommon issues:")
        print("  - Check your PRIVATE_KEY in .env")
        print("  - Ensure you have USDC in your wallet")
        print("  - Verify your internet connection")
        return 1


if __name__ == "__main__":
    sys.exit(main())

File: internal/cli/test_connection.go
package cli

import (
	"context"
	"fmt"
	"time"

	"github.com/spf13/cobra"

	"limitorderbot/internal/chain"
	"limitorderbot/internal/clob"
	"limitorderbot/internal/config"
	"limitorderbot/internal/gamma"
)

func newTestConnectionCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "test-connection",
		Short: "测试 Gamma/CLOB/RPC 连接",
		RunE: func(cmd *cobra.Command, args []string) error {
			cfg, err := config.Load()
			if err != nil {
				return err
			}

			fmt.Println("\n" + repeat("=", 60))
			fmt.Println("CONFIGURATION TEST")
			fmt.Println(repeat("=", 60))
			fmt.Println("[OK] Configuration loaded successfully")
			fmt.Printf("  - Chain ID: %d\n", cfg.ChainID)
			fmt.Printf("  - Signature Type: %s\n", cfg.SignatureType)
			fmt.Printf("  - Order Size: $%.2f\n", cfg.OrderSizeUSD)
			fmt.Printf("  - Spread Offset: %.4f\n", cfg.SpreadOffset)
			fmt.Printf("  - Check Interval: %ds\n", cfg.CheckIntervalSeconds)

			fmt.Println("\n" + repeat("=", 60))
			fmt.Println("GAMMA API TEST")
			fmt.Println(repeat("=", 60))
			disc := gamma.New(cfg.GammaAPIBaseURL)
			ctx, cancel := context.WithTimeout(context.Background(), 20*time.Second)
			defer cancel()
			markets, err := disc.DiscoverBTC15mMarkets(ctx)
			if err != nil {
				return fmt.Errorf("[FAIL] Gamma API error: %w", err)
			}
			fmt.Println("[OK] Successfully connected to Gamma API")
			fmt.Printf("  - Found %d BTC 15m markets\n", len(markets))
			for i := 0; i < len(markets) && i < 3; i++ {
				fmt.Printf("    - %s\n", markets[i].MarketSlug)
				fmt.Printf("      Start: %s\n", markets[i].StartTime().Format(time.RFC3339))
			}

			fmt.Println("\n" + repeat("=", 60))
			fmt.Println("CLOB CLIENT TEST")
			fmt.Println(repeat("=", 60))
			cc, err := clob.NewClient(cfg.ClobAPIURL, cfg.ChainID, cfg.PrivateKey, cfg.SignatureType, cfg.FunderAddress)
			if err != nil {
				return fmt.Errorf("[FAIL] CLOB client init error: %w", err)
			}
			fmt.Printf("[OK] CLOB signer initialized\n")
			fmt.Printf("  - Wallet address: %s\n", cc.Address())

			// Derive creds (best-effort; some users run read-only)
			ctx2, cancel2 := context.WithTimeout(context.Background(), 20*time.Second)
			defer cancel2()
			creds, err := cc.CreateOrDeriveAPICreds(ctx2, 0)
			if err == nil && creds.APIKey != "" {
				cc.SetCreds(creds)
				fmt.Println("[OK] CLOB API creds derived")
			} else {
				fmt.Printf("[WARNING] Could not derive CLOB API creds (read-only OK): %v\n", err)
			}

			ch, err := chain.New(cfg.RPCURL, cfg.PrivateKey, cfg.ChainID)
			if err != nil {
				return fmt.Errorf("[FAIL] RPC client init error: %w", err)
			}
			defer ch.Close()
			ctx3, cancel3 := chain.WithTimeout(context.Background(), 20*time.Second)
			defer cancel3()
			bal, err := ch.USDCBalance(ctx3)
			if err != nil {
				return fmt.Errorf("[FAIL] USDC balance error: %w", err)
			}
			fmt.Println("[OK] Successfully connected to RPC")
			fmt.Printf("  - USDC Balance: $%.2f\n", bal)

			return nil
		},
	}
}

func repeat(s string, n int) string {
	out := ""
	for i := 0; i < n; i++ {
		out += s
	}
	return out
}
